### PR TITLE
Updated metadata schema and added tests

### PR DESCRIPTION
**Pull request by samizzy**
Internal ticket: 1084

### CODE REVIEW

File: core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+         ],
+         "default":null
+      },
+      {
+         "name":"parent_eids",
+         "type":[
+            "null",
+            {
+            "type":"array",
+            "items":"string"
```
Comment (a1exsh) [inline core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc:84]: what about `logicalType`?  it should be `uuid`, as for the `eid` field

File: core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc [hunk 2]
```diff
@@ -0,0 +1,97 @@
+{
+   "name":"Metadata",
```
Comment (a1exsh) [inline core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc:2]: The lack of space after `:` makes the whole file look a bit odd.

File: core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            }
+         ],
+         "default":null
+      },
+      {
+         "name":"schema_version",
+         "type":"string"
+      },
+      {
+         "name":"published_by",
```
Comment (a1exsh) [inline core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc:45]: I think this should also be nullable, as it's set by Nakadi, not by the clients.

File: core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java [hunk 4]
```diff
@@ -63,6 +72,89 @@ public void testDeserializeAvro() throws IOException {
...
+                Byte.parseByte(metadataVersion),
+                getMetadataWriter(metadataVersion, "1", modifyMetadata),
+                eventWriter);
+
+        // try to deserialize that data when we would read Kafka record
+        final byte[] deserializedEvent = deserializer.deserialize(
+                NakadiRecord.Format.AVRO.getFormat(),
+                data
+        );
+    }
```
Comment (a1exsh) [inline core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java:155]: Can we add any assertion here?

File: core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc [hunk 5]
```diff
@@ -0,0 +1,106 @@
+{
+  "name": "Metadata",
+  "type": "record",
+  "namespace": "org.zalando.nakadi.generated.avro",
```
Comment (samizzy) [inline core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc:4]: Added namespace as well, which will be taken as package when generating stubs

Comment (adyach) [reply core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc:4]: where usually generated classes go? do they have a separate module?

File: core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  "doc": "Event metadata defines data is not related to actual event payload data",
```
Comment (adyach) [inline core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc:5]: maybe we can say instead to what that date relates? like `Event metadata defines data about the payload and additional information for Nakadi operations`

File: core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc (hunk 5 above)
Comment (samizzy) [reply core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc:4]: They are not generated and they dont have a separate module currently. The temporary solution is user of generated classes will have a copy of the .avsc file and will generate the classes locally.

File: core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc (hunk 6 above)
Comment (samizzy) [reply core-common/src/main/resources/event-type-schema/_metadata/_metadata.3.avsc]: Alright that sounds better, updated the same

**Comment by adyach** [comment]
👍

**Comment by adyach** [comment]
👍

**Comment by samizzy** [comment]
:+1:

File: core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java [hunk 7]
```diff
@@ -52,17 +62,155 @@ public void testDeserializeAvro() throws IOException {
...
+
+        final JSONObject jsonObject = new JSONObject()
+        .put("flow_id", "hek")
+        .put("partition", 0)
+        .put("received_at", SOME_TIME)
+        .put("published_by", "nakadi-test");
+
+        final var actualJson = getSerializedJsonObject(metadataVersion, jsonObject);
+        final var expectedJson = getExpectedNode0(jsonObject.put("received_at", SOME_TIME_DATE_STRING));
+        Assert.assertTrue(expectedJson.similar(actualJson));
```
Comment (a1exsh) [inline core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java:84]: Is `similar` a strong enough test here?  Shouldn't we test for real equality?

Comment (samizzy) [reply core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java:84]: JSONObject does not ovveridde equals method, so it just checks referential equality. From docs about similar,
```
    /**
     * Determine if two JSONObjects are similar.
     * They must contain the same set of names which must be associated with
     * similar values.
     *
     * @param other The other JSONObject
     * @return true if they are equal
     */

```

Comment (a1exsh) [reply core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java:84]: I found the docs as well.  But what does "similar values" really mean?  Is it a recursive application of the same method?  What about non-object values?  Is it enough for our test (e.g. do we ignore the actual values)?

Comment (samizzy) [reply core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java:84]: Yes, its a recursive function as long as its a `JSONObject` or `JSONArray`, but when we get to actual values (base case), the `equals()` method of the value is called. For non - object values (primitives?), the equals method of Primitive Wrapper classes should be getting called. Even though its called similar, it is the strongest equality check offered, perhaps it should just be called equals, although there be some reason they decided to call it similar instead (maybe order of elements is not important in similar, thats what I understand reading the implementation code)

Comment (a1exsh) [reply core-common/src/test/java/org/zalando/nakadi/repository/kafka/KafkaRecordDeserializerTest.java:84]: OK, makes sense.  Thank you :)

**Pull request merged**